ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs

```diff
@@ -12,18 +12,91 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
   , ClientStNext (..)
   , ClientPipelinedStIntersect (..)
   , ChainSyncInstruction (..)
+  , TraceChainsyncClientObservation (..)
+  , TraceChainSyncClientReqRsp
+  , TraceChainSyncClientTag (..)
 
   , chainSyncClientPeerPipelined
   , chainSyncClientPeerSender
   , mapChainSyncClientPipelined
+  , peerReqRspTracer
   ) where
 
+import Data.Functor (void)
 import Data.Kind (Type)
-
+import Control.Monad.Class.MonadSTM.Strict
+import Control.Monad.Class.MonadTime
+import Control.Tracer (Tracer (..), traceWith)
 import Network.TypedProtocol.Core
 import Network.TypedProtocol.Pipelined
 
 import Ouroboros.Network.Protocol.ChainSync.Type
+import Ouroboros.Network.Util.TaggedObserve
+
+
+-- | TraceChainsyncClientObservation indicates which request-response exchange
+-- we're observing, either MsgFindIntersect or a non-pipelined MsgRequestNext.
+data TraceChainsyncClientObservation = IntersectObservation ObserveIndicator
+                                     | NextObservation ObserveIndicator
+                                     deriving Show
+
+
+data TraceChainSyncClientTag peer = IntersectTag peer
+                                  | NextTag peer
+
+
+-- | TraceChainSyncClientReqRsp traces the start, end and duration of
+-- TraceChainSyncClientTag:ed observations.
+type TraceChainSyncClientReqRsp peer =
+    TaggedObservable (TraceChainSyncClientTag peer) Time Time DiffTime
+
+
+instance (Show peer) => Show (TraceChainSyncClientTag peer) where
+    show (IntersectTag peer) = "IntersectTag " ++ show peer
+    show (NextTag peer)      = "NextTag " ++ show peer
```

Comment:
You can just derive this instance, a few lines above where 'TraceChainSyncClientTag` is introduced.

Reply:
I don't so how.

Reply:
```
data TraceChainSyncClientTag peer = IntersectTag peer
                                  | NextTag peer
  deriving Show
```